lang/java/compiler/src/main/java/org/apache/avro/compiler/idl/SchemaResolver.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2015 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.avro.compiler.idl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.avro.JsonProperties;
+import org.apache.avro.LogicalType;
+import org.apache.avro.Protocol;
+import org.apache.avro.Schema;
+import org.codehaus.jackson.JsonNode;
+
+/**
+ * Utility class to resolve schemas that are unavailable at the time they are referenced in the IDL.
+ */
+
+final class SchemaResolver {
```

Comment:
I agree with that principle for the public API, but I treat internal classes differently. I tend to avoid things that, if we extend the class, will cause needless changes later. Either way, if I choose to extend this class later I will have to learn how it works and avoid pitfalls, so having to remove that it is final seems like more work.